runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/GrpcWindmillServer.java

```diff
@@ -987,6 +998,84 @@ public void append(StreamingGetWorkResponseChunk chunk) {
 
         this.data = data.concat(chunk.getSerializedWorkItem());
         this.bufferedSize += chunk.getSerializedWorkItem().size();
+        for (GetWorkStreamTimingInfo info : chunk.getPerWorkItemTimingInfosList()) {
+          getWorkStreamTimings.compute(
+              info.getEvent(),
+              (event, recordedTime) -> {
+                Instant newTimingForEvent = Instant.ofEpochMilli(info.getTimestampUsec() / 1000);
```

Comment:
add comment on current strategy
- take min across chunks for each event and use that to divide up transit time

As mentioned offline it seems tricky because:
- we want total time for all chunks for the work item to match real time elapsed so it aligns with other times tracked for the work item.
- chunks may be delayed due to downstream pushback

Another possible way to track would be identifying the max elapsed time within each stage and scaling.

Reply:
Added comments. I agree that we probably want to identifying the max elapsed time with each stage, which at least tells us where exactly the slowness is. I've changed implementation to take that.